core/comp_ui.py

```diff
@@ -30,18 +30,29 @@
 PROMPT_UNDERLINE = '\x01%s\x02' % _UNDERLINE
 PROMPT_REVERSE = '\x01%s\x02' % _REVERSE
 
-
 class PromptState(object):
   """For the InteractiveLineReader to communicate with the Display callback."""
 
   def __init__(self):
     self.last_prompt_str = None
     self.last_prompt_len = -1
 
+  def _PromptLen(self, prompt_str):
```

Comment:
Please make this a free function, since it doesn't use `self`.  (pet peeve of mine)

Also even though it's simple, I would write a few unit tests ... i.e. testing the case where they're adjacent, mismatched, etc.

The `elif escaped` makes me skip a beat although it might happen to be correct.


As far as unmatched \[ \], that's a good point.  Maybe in prompt.py we should render an error string like we do when `\W` can't be determined.